source/common/config/unified_mux/sotw_subscription_state.cc

```diff
@@ -0,0 +1,126 @@
+  std::vector<envoy::service::discovery::v3::Resource> resources_with_ttl(
+      message.resources().size());
+
+  {
+    const auto scoped_update = ttl_.scopedTtlUpdate();
+    for (const auto& any : message.resources()) {
+      if (!any.Is<envoy::service::discovery::v3::Resource>() &&
+          any.type_url() != message.type_url()) {
+        throw EnvoyException(fmt::format("type URL {} embedded in an individual Any does not match "
+                                         "the message-wide type URL {} in DiscoveryResponse {}",
+                                         any.type_url(), message.type_url(),
+                                         message.DebugString()));
+      }
+
+      auto decoded_resource =
+          DecodedResourceImpl::fromResource(resource_decoder_, any, message.version_info());
+      setResourceTtl(*decoded_resource);
+      if (isHeartbeatResource(*decoded_resource, message.version_info())) {
+        continue;
+      }
+      non_heartbeat_resources.Add()->CopyFrom(any);
+    }
+  }
+
+  // TODO (dmitri-d) to eliminate decoding of resources twice consider expanding the interface to
+  // support passing of decoded resources
+  callbacks().onConfigUpdate(non_heartbeat_resources, message.version_info());
+  // Now that we're passed onConfigUpdate() without an exception thrown, we know we're good.
+  last_good_version_info_ = message.version_info();
+  last_good_nonce_ = message.nonce();
+  ENVOY_LOG(debug, "Config update for {} (version {}) accepted with {} resources", typeUrl(),
+            message.version_info(), message.resources().size());
+}
+
+std::unique_ptr<envoy::service::discovery::v3::DiscoveryRequest>
+SotwSubscriptionState::getNextRequestInternal() {
+  auto request = std::make_unique<envoy::service::discovery::v3::DiscoveryRequest>();
+  request->set_type_url(typeUrl());
+  std::copy(names_tracked_.begin(), names_tracked_.end(),
+            Protobuf::RepeatedFieldBackInserter(request->mutable_resource_names()));
+  if (last_good_version_info_.has_value()) {
+    request->set_version_info(last_good_version_info_.value());
+  }
+  // Default response_nonce to the last known good one. If we are being called by
+  // getNextRequestWithAck(), this value will be overwritten.
+  if (last_good_nonce_.has_value()) {
+    request->set_response_nonce(last_good_nonce_.value());
+  }
+
+  update_pending_ = false;
+  return request;
+}
+
+void SotwSubscriptionState::ttlExpiryCallback(const std::vector<std::string>& expired) {
+  Protobuf::RepeatedPtrField<std::string> removed_resources;
+  for (const auto& resource : expired) {
+    removed_resources.Add(std::string(resource));
+  }
+  callbacks().onConfigUpdate({}, removed_resources, "");
+}
+
+void SotwSubscriptionState::setResourceTtl(const DecodedResourceImpl& decoded_resource) {
+  if (decoded_resource.ttl()) {
+    ttl_.add(std::chrono::milliseconds(*decoded_resource.ttl()), decoded_resource.name());
```

Comment:
We're setting the global TTL for the subscription state to a specific resource's TTL?

Reply:
`ttl_` is an instance of a `TtlManager`, `add` updates the ttl for a resource with a specific name (second parameter). Also: not sure what global TTL is...

Reply:
Got it, ignore this comment.